File: Low Level Design (LLD)/DesignPatterns/3. Structural/State/Java/Document.java
public class Document {
    private State state;

    public Document() {
        state = new DraftState(); // Initial state
    }

    public void setState(State state) {
        this.state = state;
    }

    public void edit() {
        state.edit(this);
    }

    public void publish() {
        state.publish(this);
    }
}

File: Low Level Design (LLD)/DesignPatterns/3. Structural/State/Cpp/src/Document.cpp
#include "../include/Document.h"
#include "../include/DraftState.h"

Document::Document() {
    state = new DraftState();
}

Document::~Document() {
    delete state;
}

void Document::setState(State* newState) {
    delete state;
    state = newState;
}

void Document::edit() {
    state->edit(this);
}

void Document::publish() {
    state->publish(this);
}
